### PR TITLE
Footer language

### PR DESCRIPTION
**Pull request by ZahraMousavi**
note: I had problem with proxy, so I used full address: `https://www.kb.dk${language === 'en' ? "/en" : ""}/jsonapi/node/site/e065d5e7-a348-4384-9859-c17841d03019?fields[node--site]=footer_column_${column}`
But have created an issue in Jira to fix it.

### CODE REVIEW

**Review by jegsen — COMMENTED** [review]
Jeg har desværre ikke så meget at komme med. Det er svært at kommentere på de enkelte ændringer for mig - også selvom jeg har kigget på projektet som et hele. Det er tydeligt at du har været meget nede i de forskellige elementer.

File: kb-fonts/package.json [hunk 1]
```diff
@@ -14,14 +26,11 @@
     "kb",
     "fonts"
   ],
-
-  "dependencies": {
-    "lit": "^2.6.1"
-  },
-  "author": "",
+  "author": "Royal Danish library",
   "license": "ISC",
```
Comment (jorntx) [inline kb-fonts/package.json:30]: Is this the correct license for us?

File: kb-footer/README.md
Comment (jorntx) [inline kb-footer/README.md:1]: Very nice with this .md documentation as we can use it directly in PL and github

File: kb-footer/package.json [hunk 2]
```diff
@@ -11,36 +14,22 @@
...
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/kb-dk/kb-web-components.git"
+  },
+  "keywords": [
+    "kb",
+    "footer"
+  ],
+  "author": "Royal Danish library",
+  "license": "ISC",
```
Comment (jorntx) [inline kb-footer/package.json:30]: Correct license?

File: kb-footer/src/kb-footer-column.ts [hunk 3]
```diff
@@ -2,73 +2,49 @@ import {directive} from "lit-html/directive.js";
...
-import {defaultsColumns} from './default_column_data.js';
+import "@kb-dk/kb-icon";
+import {defaultsColumnsDA, defaultsColumnsEN, lastColumnDA, lastColumnEN} from './default_column_data.js';
 
 class KbFooterColumn extends AsyncDirective {
 
-    update = (part:Part, [column]: DirectiveParameters<this>): void => {
-        let footerColumnDataUrl: string = `/jsonapi/node/site/e065d5e7-a348-4384-9859-c17841d03019?fields[node--site]=footer_column_${column}`;
-        this.fetchData(footerColumnDataUrl, column);
+    update = (part:Part, [column, language]: DirectiveParameters<this>): void => {
```
Comment (jorntx) [inline kb-footer/src/kb-footer-column.ts:10]: Go with 'const' instead of 'let' here

File: kb-footer/src/kb-footer-column.ts [hunk 4]
```diff
@@ -2,73 +2,49 @@ import {directive} from "lit-html/directive.js";
...
+                let defaultsColumns = language === 'en' ? defaultsColumnsEN : defaultsColumnsDA;
+                this.setValue(this.getHtml(defaultsColumns.data.attributes[`footer_column_${column}`], column, language));
+                column === 1 ? this.getAppFooterColumn(language) : '';
             });
     };
 
-    getAppFooterColumn = () : void => {
-        let ul = document.querySelector('#appFooterColumn');
+    getAppFooterColumn = (language: string) : void => {
+        let ul = this.getAppColumn(language);
```
Comment (jorntx) [inline kb-footer/src/kb-footer-column.ts]: Go with 'const' instead of 'let' here

File: kb-footer/src/kb-footer-column.ts [hunk 5]
```diff
@@ -2,73 +2,49 @@ import {directive} from "lit-html/directive.js";
...
         if (ul){
-            let shadowRoot = document.querySelector('kb-footer')?.shadowRoot;
-            shadowRoot?.querySelector('.col-sm-6.col-lg-3 ul')?.remove();
-            shadowRoot?.querySelector('.col-sm-6.col-lg-3')?.append(ul);
+            let clonedUl = ul.cloneNode(true);
```
Comment (jorntx) [inline kb-footer/src/kb-footer-column.ts]: Go with 'const' instead of 'let' here

File: kb-footer/src/kb-footer-column.ts [hunk 6]
```diff
@@ -2,73 +2,49 @@ import {directive} from "lit-html/directive.js";
...
+        let shadowRoot = document.querySelector('kb-footer')?.shadowRoot;
+        shadowRoot?.querySelector('.col-sm-6.col-lg-3 ul')?.remove();
+        shadowRoot?.querySelector('.col-sm-6.col-lg-3')?.append(clonedUl);
+    }
+
+    private getAppColumn(language: string) {
+        let columnId = language === 'en' ? 'appFooterColumnEN' : 'appFooterColumnDA';
+        let ul = document.querySelector(`#${CSS.escape(columnId)}`);
+        return ul ? ul : document.querySelector('#appFooterColumn');
+    }
```
Comment (jorntx) [inline kb-footer/src/kb-footer-column.ts:47]: Very nice clean up you did here!

File: kb-footer/src/kb-footer-column.ts [hunk 7]
```diff
@@ -78,23 +54,26 @@ class KbFooterColumn extends AsyncDirective {
             <h2 class="h3" id="${this.ariaLabels[column]}">${listData[0]?.title}</h2>
             <ul>
                 ${listData.slice(1).map(itemData => html`
-                    <li><a href="${this.getUri(itemData.uri)}">${itemData.title.includes(":cookie:") ? itemData.title.substring(8) : itemData.title}</a></li>`)}
+                    <li><a href="${this.fixLink(itemData.uri)}">${itemData.title.includes(":cookie:") ? itemData.title.substring(8) : itemData.title}</a></li>`)}
             </ul>
```
Comment (jorntx) [inline kb-footer/src/kb-footer-column.ts:58]: Good hint with the comment - so we and others know what this 'hack' is about

File: kb-footer/src/kb-footer-column.ts [hunk 8]
```diff
@@ -78,23 +54,26 @@ class KbFooterColumn extends AsyncDirective {
...
     }
 
-    render = (column: number): TemplateResult => this.getHtml(defaultsColumns, column);
+    render = (column: number, language: string): TemplateResult => {
+        let defaultsColumns = language === 'en' ? defaultsColumnsEN : defaultsColumnsDA;
+        return this.getHtml(defaultsColumns, column, language);
+    }
 
-    private getUri(uri) {
+    private fixLink(uri) {
```
Comment (jorntx) [inline kb-footer/src/kb-footer-column.ts:76]: Good rename - signals something that should be changed :)

File: kb-footer/src/kb-footer-column.ts [hunk 9]
```diff
@@ -2,73 +2,49 @@ import {directive} from "lit-html/directive.js";
...
-        let footerColumnDataUrl: string = `/jsonapi/node/site/e065d5e7-a348-4384-9859-c17841d03019?fields[node--site]=footer_column_${column}`;
-        this.fetchData(footerColumnDataUrl, column);
+    update = (part:Part, [column, language]: DirectiveParameters<this>): void => {
+        let footerColumnDataUrl: string = `/footerapi${language === 'en' ? "/en" : ""}/jsonapi/node/site/e065d5e7-a348-4384-9859-c17841d03019?fields[node--site]=footer_column_${column}`;
+        this.fetchData(footerColumnDataUrl, column, language);
     };
 
-    fetchData = (url: string, column: number): void => {
+    fetchData = (url: string, column: number, language: string): void => {
         fetch(url)
```
Comment (jorntx) [inline kb-footer/src/kb-footer-column.ts:16]: I know it is implicit when working with web components but I still like the use of the native fetch API for XHR:)

File: kb-footer/src/kb-footer.ts [hunk 10]
```diff
@@ -1,17 +1,18 @@
...
     render() {
         return html`
             <footer class="global-footer">
                 <nav class="container">
                     <div class="row">
-                        ${[1, 2, 3, 4].map(i => html`${kbFooterColumn(i)}`)}
+                        ${[1, 2, 3, 4].map(i => html`${kbFooterColumn(i, this.language)}`)}
                     </div>
                 </nav>
             </footer>
```
Comment (jorntx) [inline kb-footer/src/kb-footer.ts:18]: So clean and sweet:)

File: kb-footer/tsconfig.json
Comment (jorntx) [inline kb-footer/tsconfig.json:1]: Noticed that 'sourceMap' and 'strict' were removed - was this intentional?

File: kb-footer/webpack.config.cjs
Comment (jorntx) [inline kb-footer/webpack.config.cjs:1]: I have moved lightly over all the webpack stuff as we have agreed to give it another go to tighten it further

File: kb-icon/src/kb-icons.ts [hunk 11]
```diff
@@ -1,6 +1,6 @@
 import {LitElement, html, css} from "lit";
 import {customElement, property} from "lit/decorators.js";
-import {icons} from "./icon-list";
+import {icons} from "./icon-list.js";
```
Comment (jorntx) [inline kb-icon/src/kb-icons.ts:3]: Yes, we will fix this in the new webpack conf:)

File: kb-icon/tsconfig.json
Comment (jorntx) [inline kb-icon/tsconfig.json:1]: Noticed that 'sourceMap' and 'strict' were removed - was this intentional?

File: kb-style/package.json [hunk 12]
```diff
@@ -2,22 +2,35 @@
...
     "url": "git+https://github.com/kb-dk/kb-web-components.git"
   },
   "keywords": [
     "kb",
-    "new component"
+    "style"
   ],
-  "author": "",
+  "author": "Royal Danish library",
   "license": "ISC",
```
Comment (jorntx) [inline kb-style/package.json:30]: Correct license?

File: kb-icon/package.json [hunk 13]
```diff
@@ -13,26 +16,23 @@
...
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/kb-dk/kb-web-components.git"
+  },
+  "keywords": [
+    "kb",
+    "icons"
+  ],
+  "author": "Royal Danish library",
+  "license": "ISC",
```
Comment (jorntx) [inline kb-icon/package.json:32]: Correct license?

File: kb-icon/tsconfig.json
Comment (jorntx) [inline kb-icon/tsconfig.json:1]: Noticed that 'sourceMap' and 'strict' were removed - was this intentional?

File: kb-template/package.json [hunk 14]
```diff
@@ -15,24 +18,18 @@
...
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/kb-dk/kb-web-components.git"
+  },
+  "keywords": [
+    "kb",
+    "template"
+  ],
+  "author": "Royal Danish library",
+  "license": "ISC",
```
Comment (jorntx) [inline kb-template/package.json:30]: Correct license?

File: kb-template/tsconfig.json
Comment (jorntx) [inline kb-template/tsconfig.json:1]: Noticed that 'sourceMap' and 'strict' were removed - was this intentional?

File: package.json [hunk 15]
```diff
@@ -15,7 +15,12 @@
     "lit-element": "^3.2.2",
     "lit-html": "^2.6.1",
     "@kb-dk/kb-footer": "1.3.0",
```
Comment (jorntx) [inline package.json]: Lets go with the newest version of the footer:)

File: tsconfig.json
Comment (jorntx) [inline tsconfig.json:1]: Noticed that 'sourceMap' and 'strict' were removed - was this intentional?

File: webpack/webpack.webcomponent-paths.cjs
Comment (jorntx) [inline webpack/webpack.webcomponent-paths.cjs:1]: As mentioned before - I have moved lightly over all the webpack stuff as we have agreed to give it another go to tighten it further

**Review by jorntx — COMMENTED** [review]
Very good job with the footer so far!

I have only minor comments and adjustments and many of them are the same:)

File: kb-fonts/package.json (hunk 1 above)
Comment (ZahraMousavi) [reply kb-fonts/package.json:30]: I don't know much about licenses, but this was suggested by 'npm init' and I didn't change it.
What do you suggest?

File: kb-footer/package.json (hunk 2 above)
Comment (ZahraMousavi) [reply kb-footer/package.json:30]: I don't know much about licenses, but this was suggested by 'npm init' and I didn't change it.
What do you suggest?

Comment (jorntx) [reply kb-footer/package.json:30]: OK. I thnk we have some documentation about KB's license policy (haven't found it yet...) - let's leave this for now until I dig up the policy

File: kb-template/package.json (hunk 14 above)
Comment (ZahraMousavi) [reply kb-template/package.json:30]: I don't know much about licenses, but this was suggested but 'npm init' and I didn't change it.
What do you suggest?

File: kb-footer/src/kb-footer-column.ts (hunk 6 above)
Comment (ZahraMousavi) [reply kb-footer/src/kb-footer-column.ts:47]: Thanks :-)

File: kb-footer/package.json (hunk 2 above)
Comment (jorntx) [reply kb-footer/package.json:30]: We use 'Apache License 2.0' for our open source software here on github

File: kb-footer/tsconfig.json
Comment (ZahraMousavi) [reply kb-footer/tsconfig.json:1]: No, it was not. I just copied the tsconfig.json content that Lit suggested.
Should I add these two back?

File: kb-icon/src/kb-icons.ts (hunk 11 above)
Comment (ZahraMousavi) [reply kb-icon/src/kb-icons.ts:3]: Sounds great :-)

File: package.json (hunk 15 above)
Comment (ZahraMousavi) [reply package.json]: yes

File: webpack/webpack.webcomponent-paths.cjs
Comment (ZahraMousavi) [reply webpack/webpack.webcomponent-paths.cjs:1]: Nice

**Comment by jegsen** [comment]
Stop så :-)

File: kb-footer/tsconfig.json
Comment (jorntx) [reply kb-footer/tsconfig.json:1]: I like the checks  'strict' enable but if they give any errors I'm willing to remove it and then implement the options behind strict on a case by case basis.

The source maps might come in handy when debugging.

**Comment by jorntx** [comment]
> 



> Stop så :-)

:)

**Review by jorntx — APPROVED** [review]
Looks like we are good to go!

**Pull request merged**